sceptre/plan/actions.py

```diff
@@ -1000,3 +1000,115 @@ def diff(self, stack_differ):
         :rtype: sceptre.diffing.stack_differ.StackDiff
         """
         return stack_differ.diff(self)
+
+    @add_stack_hooks
+    def detect_stack_drift(self):
+        """
+        Detects stack drift for a running stack.
+
+        :returns: The stack drift.
+        :rtype: Tuple[str, Union[str, dict]]
+        """
+        response = self._detect_stack_drift()
+
+        detection_id = response["StackDriftDetectionId"]
+        status = self._wait_for_drift(detection_id)
+
+        if status == "DETECTION_COMPLETE":
+            response = self._describe_stack_resource_drifts()
+            return_value = (self.stack.external_name, response)
+        else:
+            return_value = (self.stack.external_name, status)
+
+        return return_value
+
+    def _wait_for_drift(self, detection_id):
+        """
+        Waits for drift detection to complete.
+
+        :param detection_id: The drift detection ID.
+        :type detection_id: str
+
+        :returns: The drift status.
+        :rtype: str
+        """
+        timeout = 300
+        elapsed = 0
+
+        while True:
+            if elapsed >= timeout:
+                return "TIMED_OUT"
+
+            self.logger.debug("%s - Waiting for drift detection", self.stack.name)
+            response = self._describe_stack_drift_detection_status(detection_id)
+
+            status = response["DetectionStatus"]
+            self._print_drift_status(response)
+
+            if status == "DETECTION_IN_PROGRESS":
+                time.sleep(10)
+                elapsed += 10
+            else:
+                return status
+
+    def _print_drift_status(self, response):
+        """
+        Print the drift status while waiting for
+        drift detection to complete.
+        """
+        keys = [
+            "StackDriftDetectionId",
+            "DetectionStatus",
+            "DetectionStatusReason",
+            "StackDriftStatus"
+        ]
+
+        for key in keys:
+            if key in response:
+                self.logger.debug(
+                    "%s - %s - %s",
+                    self.stack.name,
+                    key, response[key]
+                )
+
+    def _detect_stack_drift(self):
+        """
+        Run detect_stack_drift.
+        """
+        self.logger.debug("%s - Detecting Stack Drift", self.stack.name)
+
+        return self.connection_manager.call(
+            service="cloudformation",
+            command="detect_stack_drift",
+            kwargs={
+                "StackName": self.stack.external_name
+            }
+        )
```

Comment:
What happens if the stack doesn't exist? When invoking Sceptre commands on a stack group, we always need to account for scenarios where the stack doesn't exist yet because it has yet to be deployed. Have you accounted for that anywhere?

Reply:
@jfalkenstein I am following the conventions in the rest of the project and just allowing an exception to be raised here:

```
[aws:personal1] /src $ sceptre --merge-vars --dir=iam-user-generic --var-file=data/personal1/common-env.yaml --var-file=data/personal1/iam-lakeformation-dataanalyst.yaml detect-stack-drift iam-user-generic.yaml
"An error occurred (ValidationError) when calling the DetectStackDrift operation: Stack [DatalakeUser-iam-user-generic] does not exist"
```

This is the same behaviour as other commands, e.g. `sceptre-change-set`. I think this is reasonable?

Reply:
That's not the same behavior as all other commands. The core commands of `launch` and `delete` fail gracefully when the stack doesn't exist. Also `diff` is made that way. There might be others too.

Because Sceptre operates on StackGroups that may include new stacks that have yet to be deployed, we need to account for cases where the stack doesn't exist. For example, when running `delete` on a stack group that has some deployed and some non-deployed stacks, it gracefully moves past deletions of stacks that don't exist. I think this command should too. Otherwise, users of this command will be forced to ensure everything is deployed FIRST before they can run this command. But I can imagine there being pre-deployment steps in a CI/CD Pipeline where drift detection might be desired.

For example, one might want to employ a workflow like "detect drift on stack group; parse results; If there is drift, launch it to resolve that drift".

Reply:
@jfalkenstein I see your point here. For the record, I checked, and only `launch` and `delete` check for the existence of the stack. And then the `diff` that you wrote recently. I could copy the the logic used in `launch` and `delete` of course.

Reply:
@jfalkenstein What do you think of this now?